token/core/zkatdlog/crypto/common/nym.go

```diff
@@ -12,118 +12,146 @@ import (
 	"github.com/pkg/errors"
 )
 
-// this implements signing identity
+// NYMSigner implements signing identity
+// This is a signer for the Signature of Knowledge of (SK, BF) such that NYM = P^SK*Q^BF
 type NYMSigner struct {
 	*NYMVerifier
 	SK *math.Zr
 	BF *math.Zr
 }
 
-// get verifier
+// Serialize returns NYMSigner bytes
 func (s *NYMSigner) Serialize() ([]byte, error) {
 	return s.NYMVerifier.Serialize()
 }
 
-// sign message anonymously using Schnorr signature
+// Sign produces a SOK signature for (SK, BF) such that NYM = P^SK*Q^BF
 func (s *NYMSigner) Sign(message []byte) ([]byte, error) {
+	if s.Curve == nil {
```

Comment:
I think in general, checking that the "self" object is correctly initialized doesn't add enough to justify its existence. 

Unlike inputs, this is the object's configuration. If we can reach a state where the object is misconfigured, we better find it in tests rather than check it at runtime, because in most cases, if we're in a real environment and we reached a forbidden state, this is an irrecoverable state.

Reply:
makes sense indeed. What about adding a constructor to NYMSigner and make it a private struct?